draft/flutter-driver.md

```diff
@@ -0,0 +1,179 @@
+
+To demonstrate this, let's implement a handler to navigate back to the root route of our app. This way we can ensure that every test starts from the root page of our application.
+
+The first step is to provide a `DataHandler` to Flutter Driver:
+
+```dart
+enableFlutterDriverExtension(handler: (payload) async {
+  print(payload);
+});
+```
+
+The handler will receive a String payload and can optionally return a String response.
+
+For the sake of simplicity let's use a String as payload for now:
+
+```dart
+enableFlutterDriverExtension(handler: (payload) async {
+  if(payload == "navigate_to_root") {
+    // do something smart here
+  }
+});
+```
+
+From here, we need to implement something that will allow us to navigate to the root of our app. I'm not sure if the following is necessarily the best way to do this (if you know a better way please let me know!), but it works and is relatively straightforward.
+
+We'll use a [NavigationObserver](https://api.flutter.dev/flutter/widgets/NavigatorObserver-class.html) to get a hold of the [NavigatorState](https://api.flutter.dev/flutter/widgets/NavigatorState-class.html), which we can use to push and pop routes. We need to be able to pass in a NavigationObserver from our test entry point,  so we can access it when we receive a command to navigate to root.
+
+Change the `main` function of your app as follows:
+
+```dart
+void main() {
+  _main(null);
+}
+
+void mainTest(NavigatorObserver navigatorObserver) {
+  _main(navigatorObserver);
+}
+
+void _main(NavigatorObserver navigatorObserver) {
+  runApp(MyApp(
+    navigatorObserver: navigatorObserver,
+  ));
+}
+
+class MyApp extends StatelessWidget {
+  final NavigatorObserver navigatorObserver;
+
+  const MyApp({Key key, this.navigatorObserver}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      // all other MaterialApp initialisation here
+      navigatorObservers: navigatorObserver == null ? [] : [navigatorObserver],
+    );
+  }
+}
+```
+
+What this does is allow us to hook up a NavigationObserver from our test wrapper like so:
+
+```dart
+import 'package:flutter/material.dart';
+import 'package:flutter_driver/driver_extension.dart';
+import 'package:my_app/main.dart' as app;
+
+void main() {
+  final navigationObserver = NavigatorObserver();
+
+  enableFlutterDriverExtension(handler: (payload) async {
+    if (payload == "navigate_to_root") {
+      navigationObserver.navigator.popUntil(ModalRoute.withName('/'));
+    }
+
+    return null;
+  });
+
+  app.mainTest(navigationObserver);
+}
+
+```
+
+Now from our tests we can send our custom command, for example in a `setUp` hook:
+
+```dart
+void main() {
+  FlutterDriver driver;
+
+  setUpAll(() async {
+    driver = await FlutterDriver.connect();
+    TestHelper.driver = driver;
+  });
+
+  tearDownAll(() async {
+    driver?.close();
```

Comment:
Maybe good to explain the ? here?

Reply:
https://www.woolha.com/tutorials/dart-using-null-aware-operators-examples

Reply:
It's kind of besides the point of the post to explain Dart language constructs. So for brevity I rather not go into this.